File: daemon/src/mx/mx_bus.py
# daemon/src/mx/mx_bus.py
import asyncio
import logging
from collections import defaultdict
from .mx_message import MxMessage
from .mx_consumer import MxConsumer
from .mx_queue import MxQueue

log = logging.getLogger("mx.bus")

class MxBus:
    def __init__(self):
        self._subscriptions: dict[str, list[tuple[MxConsumer, MxQueue]]] = defaultdict(list)

    def subscribe(self, subject: str, consumer: MxConsumer, queue: MxQueue):
        self._subscriptions[subject].append((consumer, queue))
        log.info(f"subscribe: {subject} -> {type(consumer).__name__}")

    def unsubscribe(self, subject: str, consumer: MxConsumer):
        if subject in self._subscriptions:
            self._subscriptions[subject] = [
                (c, q) for c, q in self._subscriptions[subject] if c is not consumer
            ]

    async def publish(self, msg: MxMessage) -> int:
        """Deliver message to all subscriber queues for this subject."""
        subs = self._subscriptions.get(msg.subject, [])
        delivered = 0
        for consumer, queue in subs:
            if await queue.post(msg):
                delivered += 1
            else:
                log.warning(f"queue full: {queue.name} dropped {msg.op.name}")
        return delivered

    def subscriber_count(self, subject: str) -> int:
        return len(self._subscriptions.get(subject, []))

File: firmware/magic/lib/Mx/mx_bus.cpp
#include "mx_bus.h"

MxBus& MxBus::instance() {
    static MxBus inst;
    return inst;
}

bool MxBus::subscribe(uint16_t subject_id, MxConsumer* consumer, MxQueue* queue) {
    for (auto& sub : m_subs) {
        if (sub.subject_id == 0) {
            sub.subject_id = subject_id;
            sub.consumer = consumer;
            sub.queue = queue;
            return true;
        }
    }
    return false;
}

bool MxBus::unsubscribe(uint16_t subject_id, MxConsumer* consumer) {
    for (auto& sub : m_subs) {
        if (sub.subject_id == subject_id && sub.consumer == consumer) {
            sub.subject_id = 0;
            sub.consumer = nullptr;
            sub.queue = nullptr;
            return true;
        }
    }
    return false;
}

uint8_t MxBus::publish(const MxMessage& msg) {
    uint8_t count = 0;
    for (auto& sub : m_subs) {
        if (sub.subject_id == msg.subject_id && sub.queue) {
            if (sub.queue->post(msg)) count++;
        }
    }
    return count;
}

uint8_t MxBus::publishFromISR(const MxMessage& msg, BaseType_t* woken) {
    uint8_t count = 0;
    for (auto& sub : m_subs) {
        if (sub.subject_id == msg.subject_id && sub.queue) {
            if (sub.queue->postFromISR(msg, woken)) count++;
        }
    }
    return count;
}
